File: dagre/lib/order/sort.js
var _ = require("../lodash");
var util = require("../util");

module.exports = sort;

function sort(entries, biasRight, usePrev) {
  var parts = util.partition(entries, function(entry) {
    // NOTE: 有fixorder的也可以排
    return (_.has(entry, "fixorder") && !isNaN(entry.fixorder)) || _.has(entry, "barycenter");
  });
  var sortable = parts.lhs,
    unsortable = _.sortBy(parts.rhs, function(entry) { return -entry.i; }),
    vs = [],
    sum = 0,
    weight = 0,
    vsIndex = 0;

  sortable.sort(compareWithBias(!!biasRight, !!usePrev));

  vsIndex = consumeUnsortable(vs, unsortable, vsIndex);

  _.forEach(sortable, function (entry) {
    vsIndex += entry.vs.length;
    vs.push(entry.vs);
    sum += entry.barycenter * entry.weight;
    weight += entry.weight;
    vsIndex = consumeUnsortable(vs, unsortable, vsIndex);
  });

  var result = { vs: _.flatten(vs, true) };
  if (weight) {
    result.barycenter = sum / weight;
    result.weight = weight;
  }
  return result;
}

function consumeUnsortable(vs, unsortable, index) {
  var last;
  while (unsortable.length && (last = _.last(unsortable)).i <= index) {
    unsortable.pop();
    vs.push(last.vs);
    index++;
  }
  return index;
}

/**
 * 配置是否考虑使用之前的布局结果
 */
function compareWithBias(bias, usePrev) {
  return function(entryV, entryW) {
    // 排序的时候先判断fixorder，不行再判断重心
    if (entryV.fixorder !== undefined && entryW.fixorder !== undefined) {
      return entryV.fixorder - entryW.fixorder;
    }
    if (entryV.barycenter < entryW.barycenter) {
      return -1;
    } else if (entryV.barycenter > entryW.barycenter) {
      return 1;
    }
    // 重心相同，考虑之前排好的顺序
    if (usePrev && entryV.order !== undefined && entryW.order !== undefined) {
      if (entryV.order < entryW.order) {
        return -1;
      } else if (entryV.order > entryW.order) {
        return 1;
      }
    }

    return !bias ? entryV.i - entryW.i : entryW.i - entryV.i;
  };
}

File: packages/layout-rust/src/dagre/order/sort.rs
use crate::dagre::order::resolve_conflicts::ResolvedBaryEntry;
use crate::dagre::order::sort_subgraph::SubgraphResult;
use crate::dagre::util;
use crate::dagre::util::PartitionResponse;
use std::cmp::Ordering;

pub fn sort(entries: &Vec<ResolvedBaryEntry>, bias_right: &bool) -> SubgraphResult {
    let parts: PartitionResponse<ResolvedBaryEntry> = util::partition(
        entries,
        Box::new(|val: &ResolvedBaryEntry| -> bool { val.barycenter.is_some() }),
    );

    let mut sortable = parts.lhs.clone();
    sortable.sort_by(|e1, e2| compare_with_bias(e1, e2, bias_right));
    let mut unsortable = parts.rhs.clone();
    unsortable.sort_by(|e1, e2| e2.i.cmp(&e1.i));

    let mut vs: Vec<String> = vec![];
    let mut sum = 0.0;
    let mut weight = 0.0;
    let mut vs_index: usize = 0;

    vs_index = consume_unsortable(&mut vs, &mut sortable, &mut vs_index);
    sortable.iter().for_each(|entry| {
        vs_index += entry.vs.len();
        vs.append(entry.vs.clone().as_mut());
        let entry_weight = entry.weight.clone().unwrap_or(0.0);
        sum += entry.barycenter.clone().unwrap_or(0.0) * entry_weight.clone();
        weight += entry_weight.clone();
        vs_index = consume_unsortable(&mut vs, &mut unsortable, &mut vs_index);
    });

    let mut result = SubgraphResult::default();
    result.vs = vs;
    if weight > 0.0 {
        result.barycenter = sum / weight;
        result.weight = weight;
    }

    return result;
}

fn consume_unsortable(
    vs: &mut Vec<String>,
    unsortable: &mut Vec<ResolvedBaryEntry>,
    index: &mut usize,
) -> usize {
    if unsortable.len() == 0 {
        return index.clone();
    }
    let mut last: ResolvedBaryEntry = unsortable.get(unsortable.len() - 1).cloned().unwrap();
    while unsortable.len() > 0 && &last.i <= index {
        vs.append(last.vs.clone().as_mut());
        last = unsortable.pop().unwrap();
        *index += 1;
    }
    return index.clone();
}

fn compare_with_bias(
    entry_v: &ResolvedBaryEntry,
    entry_w: &ResolvedBaryEntry,
    bias: &bool,
) -> Ordering {
    let barycenter_v = entry_v.barycenter.clone().unwrap_or(0.0) as i32;
    let barycenter_w = entry_w.barycenter.clone().unwrap_or(0.0) as i32;
    if barycenter_v < barycenter_w {
        return Ordering::Greater;
    } else if barycenter_v > barycenter_w {
        return Ordering::Less;
    }

    return if !bias {
        entry_v.i.cmp(&entry_w.i)
    } else {
        entry_w.i.cmp(&entry_v.i)
    };
}
